exla/lib/exla/defn.ex

```diff
@@ -312,15 +312,15 @@ defmodule EXLA.Defn do
        ) do
     [initial_arg, _arg, pred, body] = args
 
+    {initial, cache} = recur_composite(initial_arg, state, cache)
+
     initial =
       if token = get_token(cache) do
```

Comment:
We need to recur on `initial_arg` first and use the cached token. Currently we get the token first, so it doesn't account for all the prior computation.